bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java

```diff
@@ -576,6 +576,14 @@ void sendReadTo(int bookieIndex, BookieId to, LedgerEntryRequest entry) throws I
         }
 
         if (isRecoveryRead) {
+            // Return BookieHandleNotAvailableException immediately if bookie is not available when recovery read.
+            // In the case of network isolation, reading a faulty bookie node will time out,
+            // which will cause a slow recovery or even timeout.
+            if (clientCtx.getBookieWatcher().isBookieUnavailable(to)) {
```

Comment:
One issue here is if the bookie is partitioned from ZK though it's effectively online and reachable from the BK client. This change will make it fail while today it's succeeding.